server/src/controllers/job_applications/addJobApplication.js

```diff
@@ -0,0 +1,199 @@
+  }
+
+  // ! application status - check if it is one of the these enums: ENUM type (Contacted, Rejected, R & I, Approved, In Progress, New)
+  // ! Note that the default is "New".
+  // ! Wait so let's say an incorrect value is given. Should we just default it to "new"?
+  // ! I think I should just return an error message if the value is not one of the enums.
+  // ! 400 status code with an error message
+
+  const validApplicationStatuses = [
+    "Contacted",
+    "Rejected",
+    "R & I",
+    "Approved",
+    "In Progress",
+    "New",
+  ];
+
+  if (!validApplicationStatuses.includes(applicationStatus)) {
+    return res
+      .status(400)
+      .json({ error: "Invalid application status provided." });
+  }
+
+  // ! check statusInCanada and see if it fits the following enums: ENUM type (Citizen, PR, refugee, student visa, open work, other)
+
+  const validStatusInCanada = [
+    "Citizen",
+    "PR",
+    "Refugee",
+    "Student Visa",
+    "Open Work",
+    "Other",
+  ];
+
+  if (!validStatusInCanada.includes(statusInCanada)) {
+    return res
+      .status(400)
+      .json({ error: "Invalid status in Canada provided." });
+  }
+
+  // ! check that customResponses is an object.
+  // ! If it is not an object, return a 400 status code with an error message.
+
+  if (typeof customResponses !== "object") {
+    return res
+      .status(400)
+      .json({ error: "Custom responses must be an object." });
+  }
+
+  // ! Confirm that the job posting is still open before creating the job application
+  // ! Check that the date of the job application is before the closing date in the job posting model.
+
+  if (jobPosting.close_date < new Date()) {
+    return res.status(400).json({ error: "Job posting is closed." });
+  }
+
+  // ! Confirm that the postalCode is valid
+
+  if (!/^[A-Za-z]\d[A-Za-z][ -]?\d[A-Za-z]\d$/.test(postalCode)) {
+    return res.status(400).json({ error: "Invalid postal code." });
+  }
+
+  // ! Confirm the phone number is valid.
+
+  if (!/^\d{10}$/.test(phone)) {
+    return res.status(400).json({ error: "Invalid phone number." });
+  }
+
+  // ! Confirm that the email is valid.
+
+  if (!/^.+@.+\..+$/.test(email)) {
+    return res.status(400).json({ error: "Invalid email." });
+  }
+
+  // create the job application, and return the created job application object
+  // use the id of the job application to set the resume field as `${id}_${name}`
+
+  // ! Create the job application in the database
+
+  const jobApplication = await JobApplication.create({
+    job_posting_id: jobPostingId,
+    name,
+    email,
+    phone,
+    postal_code: postalCode,
+    resume,
+    status_in_canada: statusInCanada,
+    application_status: applicationStatus,
+    custom_responses: customResponses,
+  });
+
+  // ! Set resume as ${id}_${name}
+  // ! I'm assuming this is the job application id? I think it is since it is unique.
+  // ! and that the string is just the link to the URL of the pdf file in the S3 bucket.
```

Comment:
So for these, I have some functions defined in this file of interacting with the S3 bucket: https://github.com/uoftblueprint/tces/blob/main/server/src/utils/s3.js. I was thinking we could use `uploadFileToS3` to upload the file within this put request. You can set the resume field for now as ${id}_${name}_${job_title} (which is also what you'd pass in for the `fileName` parameter in `uploadFileToS3`.
We could have another endpoint for actually getting the url to download the file: https://linear.app/tces/issue/TCES-48/endpoint-to-download-resume 

What do you think about that?

Reply:
I think that works for me. I'll start integrating the uploading and getting content to and from the S3 bucket.

Reply:
_test